api/src/main/java/org/pmiops/workbench/profile/ProfileService.java

```diff
@@ -286,50 +287,132 @@ private void updateInstitutionalAffiliations(Profile updatedProfile, DbUser user
-    if (field == null) {
-      throw new BadRequestException(String.format("%s cannot be left blank!", fieldName));
+  /**
+   * Returns a list of PropertyChange entries from a diff that match either a field or any of its
+   * subfields. For example, a pathPrefix of 'address' will match a field change for both 'address'
+   * and for 'address.zipCode'.
+   */
+  private List<Change> getChangesWithPrefix(final Diff diff, final String pathPrefix) {
+    return diff.getChanges(
+        change -> {
+          if (change instanceof PropertyChange) {
+            return ((PropertyChange) change).getPropertyNameWithPath().startsWith(pathPrefix);
+          } else {
+            return false;
+          }
+        });
+  }
+
+  /**
+   * Validates a set of Profile changes by comparing the updated profile to the previous version.
+   * Only fields that have changed are subject to validation.
+   *
+   * <p>If the previous version is null, the updated profile is presumed to be a new profile and all
+   * validation rules are run. If both versions are non-null, only changed fields are validated.
+   *
+   * @param updatedProfile
+   * @param prevProfile
+   * @throws BadRequestException
+   */
+  public void validateProfile(Profile updatedProfile, Profile prevProfile)
+      throws BadRequestException {
+    cleanProfile(updatedProfile);
+
+    boolean isNewObject = prevProfile == null;
+    Diff diff = javers.compare(prevProfile, updatedProfile);
+
+    if (!getChangesWithPrefix(diff, "username").isEmpty() || isNewObject) {
+      validateUsername(updatedProfile);
     }
-    if (field.length() > max) {
-      throw new BadRequestException(
-          String.format("%s length exceeds character limit. (%d)", fieldName, max));
+    if (!getChangesWithPrefix(diff, "contactEmail").isEmpty() || isNewObject) {
+      validateContactEmail(updatedProfile);
     }
-    if (field.length() < min) {
-      if (min == 1) {
-        throw new BadRequestException(String.format("%s cannot be left blank.", fieldName));
-      } else {
-        throw new BadRequestException(
-            String.format("%s is under character minimum. (%d)", fieldName, min));
+    if (!getChangesWithPrefix(diff, "givenName").isEmpty() || isNewObject) {
+      validateGivenName(updatedProfile);
+    }
+    if (!getChangesWithPrefix(diff, "familyName").isEmpty() || isNewObject) {
+      validateFamilyName(updatedProfile);
+    }
+    if (!getChangesWithPrefix(diff, "address").isEmpty() || isNewObject) {
+      validateAddress(updatedProfile);
+    }
+    if (!getChangesWithPrefix(diff, "areaOfResearch").isEmpty() || isNewObject) {
+      validateAreaOfResearch(updatedProfile);
+    }
+    if (!getChangesWithPrefix(diff, "verifiedInstitutionalAffiliation").isEmpty() || isNewObject) {
+      validateInstitutionalAffiliation(updatedProfile);
+    }
+
+    if (!isNewObject) {
+      // We disallow changes in certain fields.
+      if (!getChangesWithPrefix(diff, "username").isEmpty()) {
```

Comment:
Can you make a list of Strings and iterate over tit since it's the same condition really? Better yet, can you simply tell Javers not to look at these fields?

Reply:
Oh, I was thinking we only wrote fields that changed. nvm.